clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -47,3 +48,36 @@ bool Sema::checkSYCLDeviceFunction(SourceLocation Loc, FunctionDecl *Callee) {
   return DiagKind != SemaDiagnosticBuilder::K_Immediate &&
          DiagKind != SemaDiagnosticBuilder::K_ImmediateWithCallStack;
 }
+
+// The SYCL kernel's 'object type' used for diagnostics and naming/mangling is
+// the first parameter to a sycl_kernel labeled function template. In SYCL1.2.1,
+// this was passed by value, and in SYCL2020, it is passed by reference.
+static const CXXRecordDecl *
+GetSYCLKernelObjectDecl(const FunctionDecl *KernelCaller) {
+  assert(KernelCaller->getNumParams() > 0 && "Insufficient kernel parameters");
```

Comment:
Is something else verifying this before we get to the assertion?

Reply:
Yes, the sycl_kernel attribute ensures this, see handleSYCLKernelAttr, and we only call this when the caller has that attribute.